utils/gcp/object_storage.cc

```diff
@@ -0,0 +1,820 @@
+};
+
+utils::gcp::storage::client::impl::impl(std::tuple<std::string, std::string, bool, uint16_t> info, std::optional<google_credentials> c, shared_ptr<seastar::tls::certificate_credentials> certs)
+    : _endpoint(std::get<0>(info))
+    , _credentials(std::move(c))
+    , _client(std::make_unique<utils::http::dns_connection_factory>(std::get<0>(info), std::get<3>(info), std::get<2>(info), gcp_storage, certs))
+{}
+
+utils::gcp::storage::client::impl::impl(std::string_view endpoint, std::optional<google_credentials> c, shared_ptr<seastar::tls::certificate_credentials> certs)
+    : impl(rest::parse_simple_uri(endpoint.empty() ? STORAGE_APIS_URI : endpoint), std::move(c), std::move(certs))
+{}
+
+using status_type = seastar::http::reply::status_type;
+
+/**
+ * Performs a REST post/put/get with credential refresh/retry.
+ */
+future<> 
+utils::gcp::storage::client::impl::send_with_retry(const std::string& path, const std::string& scope, body_variant body, std::string_view content_type, handler_func_ex handler, httpclient::method_type op, key_values headers) {
+    int retries = 0;
+
+    for (;;) {
+        rest::request_wrapper req(_endpoint);
+        req.target(path);
+        req.method(op);
+
+        if (_credentials) {
+            try {
+                co_await _credentials->refresh(scope, &storage_scope_implies, _certs);
+                req.add_header(utils::gcp::AUTHORIZATION, format_bearer(_credentials->token));
+            } catch (...) {
+                std::throw_with_nested(permission_error("Error refreshing credentials"));
+            }
+        }
+
+        for (auto& [k,v] : headers) {
+            req.add_header(k, v);
+        }
+
+        if (!content_type.empty()) {
+            req.add_header(httpclient::CONTENT_TYPE_HEADER, content_type);
+        }
+
+        std::visit(overloaded_functor {
+            [&](const std::string& s) { req.content(s); },
+            [&](const writer_and_size& ws) { req.content(ws.first, ws.second); }
+        }, body);
+
+        // GCP storage requires this even if content is empty
+        req.add_header("Content-Length", std::to_string(req.request().content_length));
+
+        gcp_storage.trace("Sending: {}", req.request());
+
+        try {
+            co_await rest::simple_send(_client, req, [&handler](const seastar::http::reply& res, seastar::input_stream<char>& in) -> future<> {
+                gcp_storage.trace("Result: {}", res);
+                if (res._status == status_type::unauthorized) {
+                    throw httpd::unexpected_status_error(res._status);
+                }
+                co_await handler(res, in);
+            });
+            break;
+        } catch (httpd::unexpected_status_error& e) {
+            gcp_storage.debug("{}: Got unexpected response: {}", _endpoint, e.status());
+            if (e.status() == status_type::unauthorized && retries++ < 3) {
```

Comment:
`e.status() == status_type::unauthorized`
401 has 4 meanings and only 2 could possibly be solved with re-authenticating

Reply:
`retries++ < 3`
magic number

Reply:
In the case of OAuth, it can only have one meaning here really. If the credentials are truly invalid, not just expired, the refresh we do before/after should fail.